clang/lib/CIR/CodeGen/CIRGenModule.cpp

```diff
@@ -501,6 +506,31 @@ void CIRGenModule::setCommonAttributes(GlobalDecl GD, mlir::Operation *GV) {
   assert(!UnimplementedFeature::setCommonAttributes());
 }
 
+void CIRGenModule::replaceGlobal(mlir::cir::GlobalOp Old,
+                                 mlir::cir::GlobalOp New) {
+  assert(Old.getSymName() == New.getSymName() && "symbol names must match");
+
+  // If the types does not match, update all references to Old to the new type.
+  auto OldTy = Old.getSymType();
+  auto NewTy = New.getSymType();
+  if (OldTy != NewTy) {
+    auto OldSymUses = Old.getSymbolUses(theModule.getOperation());
+    if (OldSymUses.has_value()) {
+      for (auto Use : *OldSymUses) {
+        auto UseOp = dyn_cast<mlir::cir::GetGlobalOp>(Use.getUser());
+        assert(UseOp && "GlobalOp symbol user is not a GetGlobalOp");
+
+        auto UseOpResultValue = UseOp.getAddr();
+        UseOpResultValue.setType(
+            mlir::cir::PointerType::get(builder.getContext(), NewTy));
+      }
```

Comment:
@bcardosolopes maybe we should make `cir.array` types mutable too since they allow forward declarations. This would allow us to modify it in place instead of having to rewrite all its uses.

@Lancern no need to apply this review. It's here just for future reference.

Reply:
Sounds good, tracking in #380